File: TOMIMSV4/DAL/Mantenimientos/Motivo Devolucion/clsLnMotivo_devolucion.vb
Imports System.Data.SqlClient
Imports System.Reflection

Public Class clsLnMotivo_devolucion

    Public Shared Sub Cargar(ByRef oBeMotivo_devolucion As clsBeMotivo_devolucion, ByRef dr As DataRow)
        Try
            With oBeMotivo_devolucion
                .IdMotivoDevolucion = IIf(IsDBNull(dr.Item("IdMotivoDevolucion")), 0, dr.Item("IdMotivoDevolucion"))
                .IdEmpresa = IIf(IsDBNull(dr.Item("IdEmpresa")), 0, dr.Item("IdEmpresa"))
                .IdPropietario = IIf(IsDBNull(dr.Item("IdPropietario")), 0, dr.Item("IdPropietario"))
                .Nombre = IIf(IsDBNull(dr.Item("Nombre")), "", dr.Item("Nombre"))
                .User_agr = IIf(IsDBNull(dr.Item("user_agr")), "", dr.Item("user_agr"))
                .Fec_agr = IIf(IsDBNull(dr.Item("fec_agr")), Date.Now, dr.Item("fec_agr"))
                .User_mod = IIf(IsDBNull(dr.Item("user_mod")), "", dr.Item("user_mod"))
                .Fec_mod = IIf(IsDBNull(dr.Item("fec_mod")), Date.Now, dr.Item("fec_mod"))
                .Activo = IIf(IsDBNull(dr.Item("activo")), False, dr.Item("activo"))
                .Es_detalle = IIf(IsDBNull(dr.Item("es_detalle")), False, dr.Item("es_detalle"))
            End With
        Catch ex As Exception
            Throw New Exception("Motivo_devolucion_Cargar: " & ex.Message)
        End Try
    End Sub

    Public Shared Function Insertar(ByRef oBeMotivo_devolucion As clsBeMotivo_devolucion, Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing

        Try

            Ins.Init("motivo_devolucion")
            Ins.Add("idmotivodevolucion", "@idmotivodevolucion", DataType.Parametro)
            Ins.Add("idempresa", "@idempresa", DataType.Parametro)
            Ins.Add("idpropietario", "@idpropietario", DataType.Parametro)
            Ins.Add("nombre", "@nombre", DataType.Parametro)
            Ins.Add("user_agr", "@user_agr", DataType.Parametro)
            Ins.Add("fec_agr", "@fec_agr", DataType.Parametro)
            Ins.Add("user_mod", "@user_mod", DataType.Parametro)
            Ins.Add("fec_mod", "@fec_mod", DataType.Parametro)
            Ins.Add("activo", "@activo", DataType.Parametro)
            Ins.Add("es_detalle", "@es_detalle", DataType.Parametro)

            Dim sp As String = Ins.SQL()
            Dim cmd As New SqlCommand With {.CommandType = CommandType.Text}

            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)

            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection, pTransaction)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDMOTIVODEVOLUCION", oBeMotivo_devolucion.IdMotivoDevolucion))
            cmd.Parameters.Add(New SqlParameter("@IDEMPRESA", oBeMotivo_devolucion.IdEmpresa))
            cmd.Parameters.Add(New SqlParameter("@IDPROPIETARIO", oBeMotivo_devolucion.IdPropietario))
            cmd.Parameters.Add(New SqlParameter("@NOMBRE", oBeMotivo_devolucion.Nombre))
            cmd.Parameters.Add(New SqlParameter("@USER_AGR", oBeMotivo_devolucion.User_agr))
            cmd.Parameters.Add(New SqlParameter("@FEC_AGR", oBeMotivo_devolucion.Fec_agr))
            cmd.Parameters.Add(New SqlParameter("@USER_MOD", oBeMotivo_devolucion.User_mod))
            cmd.Parameters.Add(New SqlParameter("@FEC_MOD", oBeMotivo_devolucion.Fec_mod))
            cmd.Parameters.Add(New SqlParameter("@ACTIVO", oBeMotivo_devolucion.Activo))
            cmd.Parameters.Add(New SqlParameter("@ES_DETALLE", oBeMotivo_devolucion.Es_detalle))


            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            cmd.Dispose()

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

            oBeMotivo_devolucion.IdMotivoDevolucion = CInt(cmd.Parameters("@IDMOTIVODEVOLUCION").Value)

        Catch ex As Exception
            Throw New Exception("Motivo_devolucion_Insertar: " & ex.Message)
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            lConnection.Dispose()
        End Try

    End Function

    Public Shared Function Actualizar(ByRef oBeMotivo_devolucion As clsBeMotivo_devolucion, Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing

        Try

            Upd.Init("motivo_devolucion")
            Upd.Add("idmotivodevolucion", "@idmotivodevolucion", DataType.Parametro)
            Upd.Add("idempresa", "@idempresa", DataType.Parametro)
            Upd.Add("idpropietario", "@idpropietario", DataType.Parametro)
            Upd.Add("nombre", "@nombre", DataType.Parametro)
            Upd.Add("user_agr", "@user_agr", DataType.Parametro)
            Upd.Add("fec_agr", "@fec_agr", DataType.Parametro)
            Upd.Add("user_mod", "@user_mod", DataType.Parametro)
            Upd.Add("fec_mod", "@fec_mod", DataType.Parametro)
            Upd.Add("activo", "@activo", DataType.Parametro)
            Upd.Add("es_detalle", "@es_detalle", DataType.Parametro)
            Upd.Where("IdMotivoDevolucion = @IdMotivoDevolucion")

            Dim sp As String = Upd.SQL()

            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}


            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection, pTransaction)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDMOTIVODEVOLUCION", oBeMotivo_devolucion.IdMotivoDevolucion))
            cmd.Parameters.Add(New SqlParameter("@IDEMPRESA", oBeMotivo_devolucion.IdEmpresa))
            cmd.Parameters.Add(New SqlParameter("@IDPROPIETARIO", oBeMotivo_devolucion.IdPropietario))
            cmd.Parameters.Add(New SqlParameter("@NOMBRE", oBeMotivo_devolucion.Nombre))
            cmd.Parameters.Add(New SqlParameter("@USER_AGR", oBeMotivo_devolucion.User_agr))
            cmd.Parameters.Add(New SqlParameter("@FEC_AGR", oBeMotivo_devolucion.Fec_agr))
            cmd.Parameters.Add(New SqlParameter("@USER_MOD", oBeMotivo_devolucion.User_mod))
            cmd.Parameters.Add(New SqlParameter("@FEC_MOD", oBeMotivo_devolucion.Fec_mod))
            cmd.Parameters.Add(New SqlParameter("@ACTIVO", oBeMotivo_devolucion.Activo))
            cmd.Parameters.Add(New SqlParameter("@ES_DETALLE", oBeMotivo_devolucion.Es_detalle))


            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            cmd.Dispose()

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex As Exception
            Throw New Exception("Motivo_devolucion_Actualizar: " & ex.Message)
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            lConnection.Dispose()
        End Try

    End Function


    Public Shared Function Eliminar(ByRef oBeMotivo_devolucion As clsBeMotivo_devolucion, Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing

        Try


            Const sp As String = " Delete from Motivo_devolucion Where(IdMotivoDevolucion = @IdMotivoDevolucion)"


            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}

            If Es_Transaccion_Remota Then

                cmd = New SqlCommand(sp, pConection)
                cmd.Transaction = pTransaction
            Else

                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)


            End If


            cmd.Parameters.Add(New SqlParameter("@IDMOTIVODEVOLUCION", oBeMotivo_devolucion.IdMotivoDevolucion))


            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            cmd.Dispose()

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex As Exception
            Throw New Exception("Motivo_devolucion_Eliminar: " & ex.Message)
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            lConnection.Dispose()
        End Try
    End Function

    Public Function EliminarTodos(Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing

        Try

            Const sp As String = " Delete from Motivo_devolucion"

            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}
            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)

            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)


            End If

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            cmd.Dispose()

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex As Exception
            Throw New Exception("Motivo_devolucion_Eliminar: " & ex.Message)
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close
            lConnection.Dispose
        End Try

    End Function

    Public Function Listar() As DataTable

        Try

            Const sp As String = "SELECT * FROM Motivo_devolucion"

            Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}

            Dim dad As New SqlDataAdapter(cmd)

            Dim dt As New DataTable
            dad.Fill(dt)

            Return dt

        Catch ex As Exception
            Throw New Exception("Motivo_devolucion_Listar: " & ex.Message)
        End Try

    End Function

    Public Function Obtener(ByRef oBeMotivo_devolucion As clsBeMotivo_devolucion) As Boolean

        Try

            Const sp As String = "SELECT * FROM Motivo_devolucion" &
            " Where(IdMotivoDevolucion = @IdMotivoDevolucion)"

            Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}

            Dim dad As New SqlDataAdapter(cmd)


            dad.SelectCommand.Parameters.Add(New SqlParameter("@IDMOTIVODEVOLUCION", oBeMotivo_devolucion.IdMotivoDevolucion))

            Dim dt As New DataTable
            dad.Fill(dt)

            If dt.Rows.Count = 1 Then
                Cargar(oBeMotivo_devolucion, dt.Rows(0))
            Else
                Throw New Exception("No se pudo obtener el registro")
            End If

            Return True

        Catch ex As Exception
            Dim vMsgError As String = String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message)
            clsLnLog_error_wms.Agregar_Error(vMsgError)
            Throw ex
        End Try

    End Function

    Public Shared Function GetAll() As List(Of clsBeMotivo_devolucion)

        Try

            Dim lReturnList As New List(Of clsBeMotivo_devolucion)
            Const sp As String = "SELECT * FROM Motivo_devolucion"
            Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}
            Dim dad As New SqlDataAdapter(cmd)
            Dim dt As New DataTable

            dad.Fill(dt)

            Dim vBeMotivo_devolucion As New clsBeMotivo_devolucion

            For Each dr As DataRow In dt.Rows

                vBeMotivo_devolucion = New clsBeMotivo_devolucion
                Cargar(vBeMotivo_devolucion, dr)
                lReturnList.Add(vBeMotivo_devolucion)

            Next

            lConnection.Dispose()
            cmd.Dispose()

            Return lReturnList

        Catch ex As Exception
            Throw New Exception("Motivo_devolucion_GetAll: " & ex.Message)
        End Try

    End Function

    Public Shared Function GetSingle(ByRef pBeMotivo_devolucion As clsBeMotivo_devolucion)

        Try

            Const sp As String = "SELECT * FROM Motivo_devolucion" &
            " Where(IdMotivoDevolucion = @IdMotivoDevolucion)"

            Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}
            Dim dad As New SqlDataAdapter(cmd)

            dad.SelectCommand.Parameters.Add(New SqlParameter("@IDMOTIVODEVOLUCION", pBeMotivo_devolucion.IdMotivoDevolucion))

            Dim dt As New DataTable
            dad.Fill(dt)

            If dt.Rows.Count = 1 Then
                Cargar(pBeMotivo_devolucion, dt.Rows(0))
            End If

            Return True

        Catch ex As Exception
            Dim vMsgError As String = String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message)
            clsLnLog_error_wms.Agregar_Error(vMsgError)
            Throw ex
        End Try

    End Function

End Class

File: WMS.DALCore/Datos_Maestros/clsLnMotivo_devolucion.cs
using System.Data;
using Microsoft.Data.SqlClient;
using Microsoft.Extensions.Configuration;
using Microsoft.VisualBasic.CompilerServices;
using WMSWebAPI.Be;

public class clsLnMotivo_devolucion
{

    private static clsInsert Ins = new clsInsert();
    private static clsUpdate Upd = new clsUpdate();

    public static void Cargar(ref clsBeMotivo_devolucion oBeMotivo_devolucion, DataRow dr)
    {
        int GetInt(string col) { return dr[col] is DBNull ? 0 : Convert.ToInt32(dr[col]); }
        bool GetBool(string col) { return dr[col] is DBNull ? false : Convert.ToBoolean(dr[col]); }
        string GetString(string col) { return dr[col] is DBNull ? "" : (Convert.ToString(dr[col]) ?? ""); }
        DateTime GetDate(string col) { return dr[col] is DBNull ? DateTime.Now : Convert.ToDateTime(dr[col]); }        

        try
        {
            oBeMotivo_devolucion.IdMotivoDevolucion = GetInt("IdMotivoDevolucion");
            oBeMotivo_devolucion.IdEmpresa = GetInt("IdEmpresa");
            oBeMotivo_devolucion.IdPropietario = GetInt("IdPropietario");
            oBeMotivo_devolucion.Nombre = GetString("Nombre");
            oBeMotivo_devolucion.User_agr = GetString("user_agr");
            oBeMotivo_devolucion.Fec_agr = GetDate("fec_agr");
            oBeMotivo_devolucion.User_mod = GetString("user_mod");
            oBeMotivo_devolucion.Fec_mod = GetDate("fec_mod");
            oBeMotivo_devolucion.Activo = GetBool("activo");
            oBeMotivo_devolucion.Es_detalle = GetBool("es_detalle");
        }
        catch (Exception ex)
        {            
            string vMsgError = ex.Message;
            throw new Exception(vMsgError);
        }
    }

    public static int Insertar(clsBeMotivo_devolucion oBeMotivo_devolucion, SqlConnection pConection, SqlTransaction pTransaction)
    {
        if (oBeMotivo_devolucion == null)
            throw new ArgumentNullException(nameof(oBeMotivo_devolucion));

        if (pConection == null)
            throw new ArgumentNullException(nameof(pConection));

        if (pTransaction == null)
            throw new ArgumentNullException(nameof(pTransaction));

        int rowsAffected = 0;

        try
        {
            Ins.Init("motivo_devolucion");
            Ins.Add("idmotivodevolucion", "@idmotivodevolucion", "F");
            Ins.Add("idempresa", "@idempresa", "F");
            Ins.Add("idpropietario", "@idpropietario", "F");
            Ins.Add("nombre", "@nombre", "F");
            Ins.Add("user_agr", "@user_agr", "F");
            Ins.Add("fec_agr", "@fec_agr", "F");
            Ins.Add("user_mod", "@user_mod", "F");
            Ins.Add("fec_mod", "@fec_mod", "F");
            Ins.Add("activo", "@activo", "F");
            Ins.Add("es_detalle", "@es_detalle", "F");

            string sp = Ins.SQL();

            using (var cmd = new SqlCommand(sp, pConection, pTransaction))
            {
                cmd.CommandType = CommandType.Text;

                cmd.Parameters.Add(new SqlParameter("@IdMotivoDevolucion", oBeMotivo_devolucion.IdMotivoDevolucion));
                cmd.Parameters.Add(new SqlParameter("@IdEmpresa", oBeMotivo_devolucion.IdEmpresa));
                cmd.Parameters.Add(new SqlParameter("@IdPropietario", oBeMotivo_devolucion.IdPropietario));
                cmd.Parameters.Add(new SqlParameter("@Nombre", oBeMotivo_devolucion.Nombre));
                cmd.Parameters.Add(new SqlParameter("@user_agr", oBeMotivo_devolucion.User_agr));
                cmd.Parameters.Add(new SqlParameter("@fec_agr", oBeMotivo_devolucion.Fec_agr));
                cmd.Parameters.Add(new SqlParameter("@user_mod", oBeMotivo_devolucion.User_mod));
                cmd.Parameters.Add(new SqlParameter("@fec_mod", oBeMotivo_devolucion.Fec_mod));
                cmd.Parameters.Add(new SqlParameter("@activo", oBeMotivo_devolucion.Activo));
                cmd.Parameters.Add(new SqlParameter("@es_detalle", oBeMotivo_devolucion.Es_detalle));

                rowsAffected = cmd.ExecuteNonQuery();
            }

            return rowsAffected;
        }
        catch (SqlException ex)
        {
            throw new Exception(ex.Message);
        }
    }

    public static int Insertar(IConfiguration config, clsBeMotivo_devolucion oBeMotivo_devolucion)
    {

        int rowsAffected = 0;
        SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? lTransaction = null;

        try
        {
            Ins.Init("motivo_devolucion");
            Ins.Add("idmotivodevolucion", "@idmotivodevolucion", "F");
            Ins.Add("idempresa", "@idempresa", "F");
            Ins.Add("idpropietario", "@idpropietario", "F");
            Ins.Add("nombre", "@nombre", "F");
            Ins.Add("user_agr", "@user_agr", "F");
            Ins.Add("fec_agr", "@fec_agr", "F");
            Ins.Add("user_mod", "@user_mod", "F");
            Ins.Add("fec_mod", "@fec_mod", "F");
            Ins.Add("activo", "@activo", "F");
            Ins.Add("es_detalle", "@es_detalle", "F");

            string sp = Ins.SQL();

            SqlCommand cmd = new SqlCommand() { CommandType = CommandType.Text };

            lConnection.Open(); lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted);
            cmd = new SqlCommand(sp, lConnection, lTransaction);

            cmd.Parameters.Add(new SqlParameter("@IdMotivoDevolucion", oBeMotivo_devolucion.IdMotivoDevolucion));
            cmd.Parameters.Add(new SqlParameter("@IdEmpresa", oBeMotivo_devolucion.IdEmpresa));
            cmd.Parameters.Add(new SqlParameter("@IdPropietario", oBeMotivo_devolucion.IdPropietario));
            cmd.Parameters.Add(new SqlParameter("@Nombre", oBeMotivo_devolucion.Nombre));
            cmd.Parameters.Add(new SqlParameter("@user_agr", oBeMotivo_devolucion.User_agr));
            cmd.Parameters.Add(new SqlParameter("@fec_agr", oBeMotivo_devolucion.Fec_agr));
            cmd.Parameters.Add(new SqlParameter("@user_mod", oBeMotivo_devolucion.User_mod));
            cmd.Parameters.Add(new SqlParameter("@fec_mod", oBeMotivo_devolucion.Fec_mod));
            cmd.Parameters.Add(new SqlParameter("@activo", oBeMotivo_devolucion.Activo));
            cmd.Parameters.Add(new SqlParameter("@es_detalle", oBeMotivo_devolucion.Es_detalle));

            rowsAffected = cmd.ExecuteNonQuery();

            if (lTransaction != null)
                lTransaction.Commit();

        }
        catch (SqlException ex1)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();
            throw new Exception(ex1.Message);
        }
        finally
        {
            if (lConnection.State == ConnectionState.Open) lConnection.Close();
            if (lConnection != null) lConnection.Dispose();
            if (lTransaction != null) lTransaction.Dispose();
        }
        return rowsAffected;
    }

    public static int Actualizar(clsBeMotivo_devolucion oBeMotivo_devolucion, SqlConnection pConection, SqlTransaction pTransaction)
    {
        if (oBeMotivo_devolucion == null)
            throw new ArgumentNullException(nameof(oBeMotivo_devolucion));

        if (pConection == null)
            throw new ArgumentNullException(nameof(pConection));

        if (pTransaction == null)
            throw new ArgumentNullException(nameof(pTransaction));

        int rowsAffected = 0;

        try
        {
            Upd.Init("motivo_devolucion");
            Upd.Add("idmotivodevolucion", "@idmotivodevolucion", "F");
            Upd.Add("idempresa", "@idempresa", "F");
            Upd.Add("idpropietario", "@idpropietario", "F");
            Upd.Add("nombre", "@nombre", "F");
            Upd.Add("user_agr", "@user_agr", "F");
            Upd.Add("fec_agr", "@fec_agr", "F");
            Upd.Add("user_mod", "@user_mod", "F");
            Upd.Add("fec_mod", "@fec_mod", "F");
            Upd.Add("activo", "@activo", "F");
            Upd.Add("es_detalle", "@es_detalle", "F");
            Upd.Where("IdMotivoDevolucion = @IdMotivoDevolucion");

            string sp = Upd.SQL();

            using (var cmd = new SqlCommand(sp, pConection, pTransaction))
            {
                cmd.CommandType = CommandType.Text;

                cmd.Parameters.Add(new SqlParameter("@IdMotivoDevolucion", oBeMotivo_devolucion.IdMotivoDevolucion));
                cmd.Parameters.Add(new SqlParameter("@IdEmpresa", oBeMotivo_devolucion.IdEmpresa));
                cmd.Parameters.Add(new SqlParameter("@IdPropietario", oBeMotivo_devolucion.IdPropietario));
                cmd.Parameters.Add(new SqlParameter("@Nombre", oBeMotivo_devolucion.Nombre));
                cmd.Parameters.Add(new SqlParameter("@user_agr", oBeMotivo_devolucion.User_agr));
                cmd.Parameters.Add(new SqlParameter("@fec_agr", oBeMotivo_devolucion.Fec_agr));
                cmd.Parameters.Add(new SqlParameter("@user_mod", oBeMotivo_devolucion.User_mod));
                cmd.Parameters.Add(new SqlParameter("@fec_mod", oBeMotivo_devolucion.Fec_mod));
                cmd.Parameters.Add(new SqlParameter("@activo", oBeMotivo_devolucion.Activo));
                cmd.Parameters.Add(new SqlParameter("@es_detalle", oBeMotivo_devolucion.Es_detalle));

                rowsAffected = cmd.ExecuteNonQuery();
            }

            return rowsAffected;
        }
        catch (SqlException ex)
        {
            throw new Exception(ex.Message);
        }
    }

    public int Eliminar(IConfiguration config, clsBeMotivo_devolucion oBeMotivo_devolucion, SqlConnection? pConection = null, SqlTransaction? pTransaction = null)
    {

        SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? lTransaction = null;

        try
        {
            const string sp = (" Delete from Motivo_devolucion" +
             "  Where(IdMotivoDevolucion = @IdMotivoDevolucion)");

            bool Es_Transaccion_Remota = (pConection != null && pTransaction != null);

            SqlCommand cmd = new SqlCommand() { CommandType = CommandType.Text };

            if (Es_Transaccion_Remota)
            {
                cmd = new SqlCommand(sp, pConection, pTransaction);
            }
            else
            {
                lConnection.Open(); lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted);
                cmd = new SqlCommand(sp, lConnection, lTransaction);
            }

            cmd.Parameters.Add(new SqlParameter("@IdMotivoDevolucion", oBeMotivo_devolucion.IdMotivoDevolucion));

            int rowsAffected = cmd.ExecuteNonQuery();

            if (!Es_Transaccion_Remota)
                if (lTransaction != null)
                    lTransaction.Commit();

            return rowsAffected;

        }
        catch (SqlException ex1)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();            
            throw new Exception(ex1.Message);
        }
        finally
        {
            if (lConnection.State == ConnectionState.Open) lConnection.Close();
            if (lConnection != null) lConnection.Dispose();
            if (lTransaction != null) lTransaction.Dispose();
        }
    }    

    public static bool GetSingle(IConfiguration config, ref clsBeMotivo_devolucion pBeMotivo_devolucion)
    {

        SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? lTransaction = null;

        try
        {

            const string sp = "Select * FROM Motivo_devolucion" +
            " Where(IdMotivoDevolucion = @IdMotivoDevolucion)";

            lConnection.Open(); lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted);

            SqlCommand cmd = new SqlCommand(sp, lConnection, lTransaction) { CommandType = CommandType.Text };
            SqlDataAdapter dad = new SqlDataAdapter(cmd);

            dad.SelectCommand.Parameters.Add(new SqlParameter("@IdMotivoDevolucion", pBeMotivo_devolucion.IdMotivoDevolucion));            

            DataTable dt = new DataTable();
            dad.Fill(dt);

            lTransaction.Commit();

            if (dt.Rows.Count == 1)
            {
                DataRow r;
                r = dt.Rows[0];
                Cargar(ref pBeMotivo_devolucion, r);
                return true;
            }

        }
        catch (SqlException ex1)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();
            throw new Exception(ex1.Message);
        }
        finally
        {
            if (lConnection.State == ConnectionState.Open) lConnection.Close();
            if (lConnection != null) lConnection.Dispose();
            if (lTransaction != null) lTransaction.Dispose();
        }
        return false;

    }

    public static List<clsBeMotivo_devolucion> GetAll(IConfiguration config)
    {

        SqlTransaction? lTransaction = null;
        List<clsBeMotivo_devolucion> lreturnList = new List<clsBeMotivo_devolucion>();

        try
        {
            const string sp = "Select * FROM Motivo_devolucion";

            using (SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST")))
            {

                lConnection.Open();

                using (lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted))
                {
                    using (SqlDataAdapter lDTA = new SqlDataAdapter(sp, lConnection))
                    {
                        lDTA.SelectCommand.CommandType = CommandType.Text;
                        lDTA.SelectCommand.Transaction = lTransaction;
                        DataTable lDataTable = new DataTable();
                        lDTA.Fill(lDataTable);

                        clsBeMotivo_devolucion vBeMotivo_devolucion = new clsBeMotivo_devolucion();

                        foreach (DataRow dr in lDataTable.Rows)
                        {
                            vBeMotivo_devolucion = new clsBeMotivo_devolucion();
                            Cargar(ref vBeMotivo_devolucion, dr);
                            lreturnList.Add(vBeMotivo_devolucion);
                        }

                        lTransaction.Commit();
                    }

                    lConnection.Close();

                }

            }

            return lreturnList;

        }
        catch (SqlException ex1)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();            
            throw new Exception(ex1.Message);
        }
    }

    public static int MaxID(IConfiguration config)
    {

        SqlTransaction? lTransaction = null;

        try
        {

            int lMax = 0;

            const string sp = "Select ISNULL(Max(IdMotivoDevolucion),0) FROM Motivo_devolucion";

            using (SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST")))
            {
                lConnection.Open();

                using (lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted))
                {
                    using (SqlCommand lCommand = new SqlCommand(sp, lConnection) { CommandType = CommandType.Text })
                    {
                        var lreturnValue = lCommand.ExecuteScalar();
                        if (lreturnValue != DBNull.Value && lreturnValue != null)
                        {
                            lMax = int.Parse((string)lreturnValue);
                        }
                    }
                    lTransaction.Commit();
                }

                lConnection.Close();
            }

            return lMax;

        }
        catch (SqlException ex1)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();
            throw new Exception(ex1.Message);
        }
    }
    public static int MaxID(IConfiguration config, SqlConnection? pConection = null, SqlTransaction? pTransaction = null)
    {

        SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? lTransaction = null;
        int lMax = 0;
        try
        {


            const string sp = "Select ISNULL(Max(IdMotivoDevolucion),0) FROM Motivo_devolucion";

            bool Es_Transaccion_Remota = pConection is not null && pTransaction is not null;
            var cmd = new SqlCommand(sp, lConnection) { CommandType = (CommandType)Conversions.ToInteger(CommandType.Text) };
            if (Es_Transaccion_Remota)
            {
                cmd = new SqlCommand(sp, pConection, pTransaction);
            }
            else
            {
                lConnection.Open(); lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted);
                cmd = new SqlCommand(sp, lConnection, lTransaction);
            }

            var lreturnValue = cmd.ExecuteScalar();

            if (lreturnValue != DBNull.Value && lreturnValue != null)
            {
                lMax = int.Parse((string)lreturnValue);
            }

            if (!Es_Transaccion_Remota)
                if (lTransaction != null)
                    lTransaction.Commit();

            return lMax;

        }
        catch (SqlException ex1)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();            
            throw new Exception(ex1.Message);
        }
    }

    public static bool Obtener(clsBeMotivo_devolucion oBeMotivo_devolucion,
                              SqlConnection lConnection,
                              SqlTransaction lTransaction)
    {
        try
        {
            const string sp = @"SELECT * FROM Motivo_devolucion 
                       WHERE IdMotivoDevolucion = @IdMotivoDevolucion";

            using (SqlCommand cmd = new SqlCommand(sp, lConnection, lTransaction))
            {
                cmd.CommandType = CommandType.Text;
                cmd.Parameters.AddWithValue("@IDMOTIVODEVOLUCION", oBeMotivo_devolucion.IdMotivoDevolucion);

                using (SqlDataAdapter dad = new SqlDataAdapter(cmd))
                {
                    DataTable dt = new DataTable();
                    dad.Fill(dt);

                    if (dt.Rows.Count == 1)
                    {
                        Cargar(ref oBeMotivo_devolucion, dt.Rows[0]);
                        return true;
                    }
                }
            }

            return false;
        }
        catch (Exception)
        {
            throw;
        }
    }
}
